### PR TITLE
Simlify and extend Function lookup

### LINKED ISSUE #741: Setting derivatives
Setting derivatives for the staggered derivatives seems to be somewhat broken in the bout mesh:
```
  output_info.write("\tSecond      : ");
  derivs_set(options, SecondDerivTable, "second", fd2d);
  if (StaggerGrids) {
    output_info.write("\tStag. Second: ");
    derivs_set(options, SecondStagDerivTable, "second", sfd2d);
  }
```
This prevents from using C4 - as there is no C4 second derivative implementation with only 2 guard cells.
For some reason the initialisation works never the less ...

Earlier I used in the aiolos mesh (e.g. in d7e592da32e032adf1c49653da698ca77588beaa):
```
  Options *dirOption;
  Options *defOption = option->getSection("diff");
  output.write("\tSetting derivatives for direction x:\n");
  dirOption = option->getSection("ddx");

  // This file is auto-generated - do not edit!
  // Setting derivatives for ddx and FirstStag
  if (dirOption->isSet("FirstStag")) {
    dirOption->get("FirstStag", name, "");
  } else if (dirOption->isSet("First")) {
    dirOption->get("First", name, "");
  } else if (dirOption->isSet("all")) {
    dirOption->get("all", name, "");
  } else if (defOption->isSet("FirstStag")) {
    defOption->get("FirstStag", name, "");
  } else if (defOption->isSet("First")) {
    defOption->get("First", name, "");
  } else if (defOption->isSet("all")) {
    defOption->get("all", name, "");
  } else {
    name = "C2";
  }
```
which is a slightly longer, but allows to set 'default' options:
```
[mesh:diff]
all=c4
secondStag=c2
flux=U1
[mesh:ddx]
first=c2
```
As there are `3*2*4=24` possible derivatives, it comes otherwise rather tedious to set them ...

### PR DESCRIPTION
**Pull request by dschwoerer**
* Deduplicated code by templating
* Return DiffLookup and convert to appropriate type as needed
* Check %sStag first, for staggered derivatives
* Check general diff location, if nothing in ddx is specified

- [x] Resolves #741 
- [x] Checks needed
- [x] Documentation needed

### CODE REVIEW

File: manual/sphinx/user_docs/bout_options.rst [hunk 1]
```diff
@@ -272,8 +272,10 @@ and problem.
 Differencing methods
 --------------------
 
-Differencing methods are specified in three section (``[ddx]``,
-``[ddy]`` and ``[ddz]``), one for each dimension.
+Differencing methods are specified in the section (``[ddx]``,
```
Comment (johnomotani) [inline manual/sphinx/user_docs/bout_options.rst]: These sections are `[mesh:ddx]`, `[mesh:ddy]` and `[mesh:ddz]` now, would be good to update the documentation here. I guess also `[mesh:diff]`?

**Comment by dschwoerer** [comment]
Should be ready for merging :+1:

File: src/mesh/index_derivs.cxx [hunk 2]
```diff
@@ -570,91 +548,79 @@ Mesh::flux_func sfFDDX, sfFDDY, sfFDDZ;
...
-  printFuncName(method);                         // Print differential function name
-  f = lookupFluxFunc(table, method);
+  label = def;
+  if (staggerGrids) {
+    output_info.write("\tStag. %-6s: ", name.c_str());
+    for (auto &_name : {name + "stag", name}) {
+      for (auto &opts : options) {
+        if (opts->isSet(_name)) {
+          opts->get(_name, label, "");
+          goto endloop;
```
Comment (ZedThree) [inline src/mesh/index_derivs.cxx]: Is the `goto` necessary here? Could you not just do

```cpp
    opts->get(_name, label, "");
    sf = lookupFunc(stable, label);
    return;
```
Am I missing something?

**Comment by ZedThree** [comment]
I think this is good, and if anything doesn't go far enough. I've been playing about today making the various deriv tables maps, and it generally makes things even nicer than this PR already does. It also gives the added advantage of being able to register new derivatives in the user code.

This could go in now, once my question about the goto has been addressed, and I'll submit another PR turning things into maps, and we can see how that looks.

File: src/mesh/index_derivs.cxx (hunk 2 above)
Comment (dschwoerer) [reply src/mesh/index_derivs.cxx]: As it only duplicates one line of code, it should be fine ...
In general `goto`s are underrated - they can be quite useful - e.g. for breaking loops :)

File: manual/sphinx/user_docs/bout_options.rst [hunk 3]
```diff
@@ -272,8 +272,13 @@ and problem.
 Differencing methods
 --------------------
 
-Differencing methods are specified in three section (``[ddx]``,
-``[ddy]`` and ``[ddz]``), one for each dimension.
+Differencing methods are specified in the section (``[meesh:ddx]``,
```
Comment (bendudson) [inline manual/sphinx/user_docs/bout_options.rst]: typo `meesh`

**Pull request merged**